src/conditions/condition.ts

```diff
@@ -0,0 +1,63 @@
+import { z } from 'zod';
+
+import { objectEquals } from '../utils';
+
+import {
+  ContractConditionProps,
+  RpcConditionProps,
+  TimeConditionProps,
+} from './base';
+import { CompoundConditionProps } from './compound-condition';
+
+// Not using discriminated union because of inconsistent Zod types
+// Some conditions have ZodEffect types because of .refine() calls
+export type ConditionProps =
+  | RpcConditionProps
+  | TimeConditionProps
+  | ContractConditionProps
+  | CompoundConditionProps;
+
+export class Condition {
+  constructor(
+    public readonly schema: z.ZodSchema,
+    public readonly value:
+      | RpcConditionProps
+      | TimeConditionProps
+      | ContractConditionProps
+      | CompoundConditionProps
+  ) {}
+
+  public validate(override: Partial<ConditionProps> = {}): {
+    data?: ConditionProps;
+    error?: z.ZodError;
+  } {
+    const newValue = {
+      ...this.value,
+      ...override,
+    };
+    const result = this.schema.safeParse(newValue);
+    if (result.success) {
+      return { data: result.data };
+    }
+    return { error: result.error };
```

Comment:
Not sure I follow why we need to return ConditionProps or Error here.


Is the returned data dictionary used somehow, didn't see how in the tests.

Are we relying on users of the API to actively call validate? What I'm wondering is how come we don't just validate in the constructor and raise an error if validation fails. `validate` would essentially be an internal function and not expected to be called externally at all. Or am I missing part of the picture?

Reply:
I see the point of a "throwing constructor", but I'd like for the user to have a public `validate` method that doesn't throw and return the errors in an orderly manner. That would allow for programmatic feedback on the condition structure. For example, imagine someone building a visual condition builder - It would make sense for them to parse these errors and return something actionable to the caller (end-user).

Perhaps we could settle on a "throwing constructor" and have `validate` as a public static method to call in case someone wants to pre-emptively validate their condition object?